connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java

```diff
@@ -0,0 +1,578 @@
+
+    final Map<String, Object> pathTree;
+    final List<SingleFieldPath> paths;
+
+    MultiFieldPaths(List<SingleFieldPath> paths) {
+        this.paths = paths.stream().filter(Objects::nonNull).collect(Collectors.toList());
+        pathTree = buildPathTree(this.paths, 0, new HashMap<>());
+    }
+
+    public static Builder newBuilder(FieldSyntaxVersion syntaxVersion) {
+        return new Builder(syntaxVersion);
+    }
+
+    public static MultiFieldPaths of(SingleFieldPath path) {
+        return new MultiFieldPaths(Collections.singletonList(path));
+    }
+
+    public static MultiFieldPaths of(SingleFieldPath... paths) {
+        return new MultiFieldPaths(Arrays.asList(paths));
+    }
+
+    public static MultiFieldPaths of(List<SingleFieldPath> paths) {
+        return new MultiFieldPaths(paths);
+    }
+
+    public static MultiFieldPaths of(Set<String> fields, FieldSyntaxVersion syntaxVersion) {
+        return new MultiFieldPaths(fields.stream()
+                .map(f -> SingleFieldPath.of(f, syntaxVersion))
+                .collect(Collectors.toList()));
+    }
+
+    public static MultiFieldPaths of(List<String> fields, FieldSyntaxVersion syntaxVersion) {
+        return new MultiFieldPaths(fields.stream()
+                .map(f -> SingleFieldPath.of(f, syntaxVersion))
+                .collect(Collectors.toList()));
+    }
+
+    Map<String, Object> buildPathTree(List<SingleFieldPath> paths, int stepIdx, Map<String, Object> pathTree) {
+        if (paths.size() == 1) { // optimize for paths with a single member
+            SingleFieldPath path = paths.get(0);
+            if (path != null) {
+                if (path.stepAt(stepIdx + 1) == null) { // if last path step
+                    pathTree.put(path.stepAt(stepIdx), path);
+                } else {
+                    pathTree.put(path.stepAt(stepIdx),
+                            buildPathTree(paths, stepIdx + 1, new HashMap<>()));
+                }
+            }
+        } else {
+            // group paths by prefix
+            final Map<String, List<SingleFieldPath>> groups = new HashMap<>();
+            for (SingleFieldPath path : paths) {
+                if (path != null) {
+                    String step = path.stepAt(stepIdx);
+                    if (step != null) {
+                        groups.computeIfPresent(step, (s, fieldPaths) -> {
+                            for (SingleFieldPath other : fieldPaths) {
+                                // avoid overlapping paths
+                                if (!path.equals(other)
+                                        && (other.stepAt(stepIdx + 1) == null
+                                        || path.stepAt(stepIdx + 1) == null)) {
+                                    throw new IllegalArgumentException(
+                                            "Path " + other + " and " + path + " are overlapping. "
+                                                    + "Paths need to point to leaf values");
```

Comment:
oh haven't think about this one! I agree it's a valid use-case.
This will require keeping "user" on the tree. 

I have gave a try and use empty string to represent "user" on the tree:
- "user"
  - ""
  - "email"